handling binary outputs

Looks like the outputs are assumed to be strings and pytest-subprocess is appending line separators. 

```
    def test_binary_output(fake_process):
        fake_process.register_subprocess(
            ["cat", "/tmp/foo"], stdout=bytes.fromhex("000102")
        )
    
        proc = subprocess.run(["cat", "/tmp/foo"], capture_output=True)
    
>       assert proc.stdout == b"\x00\x01\x02"
E       AssertionError: assert b'\x00\x01\x02\n' == b'\x00\x01\x02'
E         Full diff:
E         - b'\x00\x01\x02'
E         + b'\x00\x01\x02\n'
E         ?               ++
```

This should check for text mode: https://github.com/aklajnert/pytest-subprocess/blob/master/pytest_subprocess/core.py#L146

--- a/README.rst
+++ b/README.rst
@@ -52,6 +52,25 @@ allows defining the fake processes behavior.
 
 .. code-block:: python
 
+    def test_echo_null_byte(fake_process):
+        fake_process.register_subprocess(
+            ["echo", "-ne", "\x00"], stdout=bytes.fromhex("00")
+        )
+
+        process = subprocess.Popen(
+            ["echo", "-ne", "\x00"],
+            stdout=subprocess.PIPE,
+        )
+        out, _ = process.communicate()
+
+        assert process.returncode == 0
+        assert out == b'\x00'
+
+Optionally, the ``stdout`` and ``stderr`` parameters can be a list (or tuple)
+of lines to be joined together with a trailing ``os.linesep`` on each line.
+
+.. code-block:: python
+
     def test_git(fake_process):
         fake_process.register_subprocess(
             ["git", "branch"], stdout=["* fake_branch", "  master"]
@@ -97,7 +116,7 @@ the input data. If the function will return a dictionary with ``stdout`` or
         process = subprocess.Popen(
             ["command"], stdin=subprocess.PIPE, stdout=subprocess.PIPE,
         )
-        out, _ = process.communicate(input=b"sample input")
+        out, _ = process.communicate(input=b"sample input\n")
 
         assert out.splitlines() == [
             b"Just stdout",
@@ -149,9 +168,9 @@ which will keep the last registered process forever.
             "test", stdout="second execution", returncode=1
         )
 
-        assert subprocess.check_output("test") == b"first execution\n"
+        assert subprocess.check_output("test") == b"first execution"
         second_process = subprocess.run("test", stdout=subprocess.PIPE)
-        assert second_process.stdout == b"second execution\n"
+        assert second_process.stdout == b"second execution"
         assert second_process.returncode == 1
 
         # 3rd time shall raise an exception
@@ -165,10 +184,10 @@ which will keep the last registered process forever.
         fake_process.keep_last_process(True)
 
         # now the processes can be called forever
-        assert subprocess.check_output("test") == b"first execution\n"
-        assert subprocess.check_output("test") == b"second execution\n"
-        assert subprocess.check_output("test") == b"second execution\n"
-        assert subprocess.check_output("test") == b"second execution\n"
+        assert subprocess.check_output("test") == b"first execution"
+        assert subprocess.check_output("test") == b"second execution"
+        assert subprocess.check_output("test") == b"second execution"
+        assert subprocess.check_output("test") == b"second execution"
 
 
 Using callbacks

--- a/pytest_subprocess/core.py
+++ b/pytest_subprocess/core.py
@@ -136,17 +136,18 @@ class FakePopen:
         if isinstance(input, (list, tuple)):
             input = linesep.join(map(self._convert, input))
 
+            # Add trailing newline if data is present.
+            if input:
+                input += linesep
+
         if isinstance(input, str) and not self.text_mode:
             input = input.encode()
 
         if isinstance(input, bytes) and self.text_mode:
             input = input.decode()
 
-        if input:
-            if not input.endswith(linesep):
-                input += linesep
-            if self.text_mode and self.__universal_newlines:
-                input = input.replace("\r\n", "\n")
+        if input and self.__universal_newlines:
+            input = input.replace("\r\n", "\n")
 
         if io_base is not None:
             input = io_base.getvalue() + (input)

--- a/tests/test_subprocess.py
+++ b/tests/test_subprocess.py
@@ -43,18 +43,18 @@ def test_multiple_levels(fake_process):
 
         assert (
             subprocess.check_output("first_command")
-            == ("first command lower-level" + os.linesep).encode()
+            == ("first command lower-level").encode()
         )
         assert (
             subprocess.check_output("second_command")
-            == ("second command lower-level" + os.linesep).encode()
+            == ("second command lower-level").encode()
         )
 
     # first command definition shall be back at top-level definition, and the second
     # command is no longer defined so it shall raise an exception
     assert (
         subprocess.check_output("first_command")
-        == ("first command top-level" + os.linesep).encode()
+        == ("first command top-level").encode()
     )
     with pytest.raises(pytest_subprocess.ProcessNotRegisteredError) as exc:
         subprocess.check_call("second_command")
@@ -135,8 +135,8 @@ def test_basic_process_merge_streams(fake_process, fake):
     if fake:
         fake_process.register_subprocess(
             ["python", "example_script.py", "stderr"],
-            stdout="Stdout line 1\nStdout line 2",
-            stderr="Stderr line 1",
+            stdout=["Stdout line 1", "Stdout line 2"],
+            stderr=["Stderr line 1"],
         )
 
     process = subprocess.Popen(
@@ -270,7 +270,8 @@ def test_universal_newlines(fake_process, fake):
     fake_process.allow_unregistered(not fake)
     if fake:
         fake_process.register_subprocess(
-            ["python", "example_script.py"], stdout=b"Stdout line 1\r\nStdout line 2",
+            ["python", "example_script.py"],
+            stdout=b"Stdout line 1\r\nStdout line 2\r\n",
         )
     process = subprocess.Popen(
         ("python", "example_script.py"), universal_newlines=True, stdout=subprocess.PIPE
@@ -301,6 +302,38 @@ def test_text(fake_process, fake):
         process.wait()
 
         assert process.stdout.read().splitlines() == ["Stdout line 1", "Stdout line 2"]
+
+
+def test_binary(fake_process):
+    fake_process.register_subprocess(
+        ["some-cmd"],
+        stdout=bytes.fromhex("aabbcc"),
+    )
+
+    process = subprocess.Popen(
+        ["some-cmd"], stdout=subprocess.PIPE
+    )
+    process.wait()
+
+    assert process.stdout.read() == b"\xaa\xbb\xcc"
+
+
+def test_empty_stdout(fake_process):
+    fake_process.register_subprocess(["some-cmd"], stdout=b"")
+
+    process = subprocess.Popen(["some-cmd"], stdout=subprocess.PIPE)
+    process.wait()
+
+    assert process.stdout.read() == b""
+
+
+def test_empty_stdout_list(fake_process):
+    fake_process.register_subprocess(["some-cmd"], stdout=[])
+
+    process = subprocess.Popen(["some-cmd"], stdout=subprocess.PIPE)
+    process.wait()
+
+    assert process.stdout.read() == b""
 
 
 @pytest.mark.parametrize("fake", [False, True])
@@ -452,10 +485,10 @@ def test_different_output(fake_process):
     # the third execution will return non-zero exit code
     fake_process.register_subprocess("test", stdout="third execution", returncode=1)
 
-    assert subprocess.check_output("test") == b"first execution" + os.linesep.encode()
-    assert subprocess.check_output("test") == b"second execution" + os.linesep.encode()
+    assert subprocess.check_output("test") == b"first execution"
+    assert subprocess.check_output("test") == b"second execution"
     third_process = subprocess.Popen("test", stdout=subprocess.PIPE)
-    assert third_process.stdout.read() == b"third execution" + os.linesep.encode()
+    assert third_process.stdout.read() == b"third execution"
     assert third_process.returncode == 1
 
     # 4-th time shall raise an exception
@@ -470,10 +503,10 @@ def test_different_output(fake_process):
     fake_process.keep_last_process(True)
 
     # now the processes can be called forever
-    assert subprocess.check_output("test") == b"first execution" + os.linesep.encode()
-    assert subprocess.check_output("test") == b"second execution" + os.linesep.encode()
-    assert subprocess.check_output("test") == b"second execution" + os.linesep.encode()
-    assert subprocess.check_output("test") == b"second execution" + os.linesep.encode()
+    assert subprocess.check_output("test") == b"first execution"
+    assert subprocess.check_output("test") == b"second execution"
+    assert subprocess.check_output("test") == b"second execution"
+    assert subprocess.check_output("test") == b"second execution"
 
 
 def test_different_output_with_context(fake_process):
@@ -488,8 +521,8 @@ def test_different_output_with_context(fake_process):
     with fake_process.context() as nested:
         nested.register_subprocess("test", stdout="nested")
 
-        assert subprocess.check_output("test") == b"nested" + os.linesep.encode()
-        assert subprocess.check_output("test") == b"top-level" + os.linesep.encode()
+        assert subprocess.check_output("test") == b"nested"
+        assert subprocess.check_output("test") == b"top-level"
 
         with pytest.raises(pytest_subprocess.ProcessNotRegisteredError) as exc:
             subprocess.check_call("test")
@@ -500,15 +533,15 @@ def test_different_output_with_context(fake_process):
         # another nest level, the top level shall reappear
         nested2.register_subprocess("test", stdout="nested2")
 
-        assert subprocess.check_output("test") == b"nested2" + os.linesep.encode()
-        assert subprocess.check_output("test") == b"top-level" + os.linesep.encode()
+        assert subprocess.check_output("test") == b"nested2"
+        assert subprocess.check_output("test") == b"top-level"
 
         with pytest.raises(pytest_subprocess.ProcessNotRegisteredError) as exc:
             subprocess.check_call("test")
 
         assert str(exc.value) == "The process 'test' was not registered."
 
-    assert subprocess.check_output("test") == b"top-level" + os.linesep.encode()
+    assert subprocess.check_output("test") == b"top-level"
 
     with pytest.raises(pytest_subprocess.ProcessNotRegisteredError) as exc:
         subprocess.check_call("test")
@@ -529,25 +562,25 @@ def test_different_output_with_context_multilevel(fake_process):
             second_level.register_subprocess("test", stdout="second-level")
 
             assert (
-                subprocess.check_output("test") == b"second-level" + os.linesep.encode()
+                subprocess.check_output("test") == b"second-level"
             )
             assert (
-                subprocess.check_output("test") == b"first-level" + os.linesep.encode()
+                subprocess.check_output("test") == b"first-level"
             )
-            assert subprocess.check_output("test") == b"top-level" + os.linesep.encode()
+            assert subprocess.check_output("test") == b"top-level"
 
             with pytest.raises(pytest_subprocess.ProcessNotRegisteredError) as exc:
                 subprocess.check_call("test")
 
-        assert subprocess.check_output("test") == b"first-level" + os.linesep.encode()
-        assert subprocess.check_output("test") == b"top-level" + os.linesep.encode()
+        assert subprocess.check_output("test") == b"first-level"
+        assert subprocess.check_output("test") == b"top-level"
 
         with pytest.raises(pytest_subprocess.ProcessNotRegisteredError) as exc:
             subprocess.check_call("test")
 
         assert str(exc.value) == "The process 'test' was not registered."
 
-    assert subprocess.check_output("test") == b"top-level" + os.linesep.encode()
+    assert subprocess.check_output("test") == b"top-level"
 
     with pytest.raises(pytest_subprocess.ProcessNotRegisteredError) as exc:
         subprocess.check_call("test")
@@ -559,17 +592,17 @@ def test_multiple_level_early_consuming(fake_process):
     be consumed before entering the context manager.
     """
     fake_process.register_subprocess("test", stdout="top-level", occurrences=2)
-    assert subprocess.check_output("test") == b"top-level" + os.linesep.encode()
+    assert subprocess.check_output("test") == b"top-level"
 
     with fake_process.context():
-        assert subprocess.check_output("test") == b"top-level" + os.linesep.encode()
+        assert subprocess.check_output("test") == b"top-level"
 
         with pytest.raises(pytest_subprocess.ProcessNotRegisteredError) as exc:
             subprocess.check_call("test")
 
         assert str(exc.value) == "The process 'test' was not registered."
 
-    assert subprocess.check_output("test") == b"top-level" + os.linesep.encode()
+    assert subprocess.check_output("test") == b"top-level"
 
     with pytest.raises(pytest_subprocess.ProcessNotRegisteredError) as exc:
         subprocess.check_call("test")
@@ -586,10 +619,10 @@ def test_keep_last_process(fake_process):
     fake_process.register_subprocess("test", stdout="First run")
     fake_process.register_subprocess("test", stdout="Second run")
 
-    assert subprocess.check_output("test") == b"First run" + os.linesep.encode()
-    assert subprocess.check_output("test") == b"Second run" + os.linesep.encode()
-    assert subprocess.check_output("test") == b"Second run" + os.linesep.encode()
-    assert subprocess.check_output("test") == b"Second run" + os.linesep.encode()
+    assert subprocess.check_output("test") == b"First run"
+    assert subprocess.check_output("test") == b"Second run"
+    assert subprocess.check_output("test") == b"Second run"
+    assert subprocess.check_output("test") == b"Second run"
 
 
 def test_git(fake_process):
@@ -609,7 +642,7 @@ def test_git(fake_process):
 def test_use_real(fake_process):
     fake_process.pass_command(["python", "example_script.py"], occurrences=3)
     fake_process.register_subprocess(
-        ["python", "example_script.py"], stdout="Fake line 1\nFake line 2"
+        ["python", "example_script.py"], stdout=["Fake line 1", "Fake line 2"]
     )
 
     for _ in range(0, 3):
@@ -707,11 +740,11 @@ def test_subprocess_pipe_without_stream_definition(fake_process):
 
     assert (
         subprocess.check_output(["test-no-stderr"], stderr=subprocess.STDOUT).decode()
-        == "test" + os.linesep
+        == "test"
     )
     assert (
         subprocess.check_output(["test-no-stdout"], stderr=subprocess.STDOUT).decode()
-        == "test" + os.linesep
+        == "test"
     )
     assert (
         subprocess.check_output(["test-no-streams"], stderr=subprocess.STDOUT).decode()

